docs/ql-libraries/dataflow/dataflow.md

```diff
@@ -0,0 +1,472 @@
+generated for calls that can be summarized by the shared library as getters.
+This usage of `PostUpdateNode`s ensures that `x` will not flow into the `getB`
+call after reaching `a`.
+
+### Example 3
+
+Consider a constructor and its call (for this example we will use Java, but the
+idea should generalize):
+```java
+MyObj(Content content) {
+  this.content = content;
+}
+
+obj = new MyObj(source);
+sink(obj.content);
+```
+
+We would like the constructor call to act in the same way as a setter, and
+indeed this is quite simple to achieve. We can introduce a synthetic data-flow
+node associated with the constructor call, let us call it `MallocNode`, and
+make this an `ArgumentNode` with position `-1` such that it hooks up with the
+implicit `this`-parameter of the constructor body. Then we can set the
+corresponding `PostUpdateNode` of the `MallocNode` to be the constructor call
+itself as this represents the value of the object after construction, that is
+after the constructor has run. With this setup of `ArgumentNode`s and
+`PostUpdateNode`s we will achieve the desired flow from `source` to `sink`
+
+### Field flow barriers
+
+Consider this field flow example:
+```
+obj.f = source;
+obj.f = safeValue;
+sink(obj.f);
+```
+or the similar case when field flow is used to model collection content:
+```
+obj.add(source);
+obj.clear();
+sink(obj.get(key));
+```
+Clearing a field or content like this should act as a barrier, and this can be
+achieved by marking the relevant `Node, Content` pair as a clear operation in
+the `clearsContent` predicate. A reasonable default implementation for fields
+looks like this:
+```ql
+predicate clearsContent(Node n, Content c) {
+  n = any(PostUpdateNode pun | storeStep(_, c, pun)).getPreUpdateNode()
+}
+```
+However, this relies on the local step relation using the smallest possible
+use-use steps. If local flow is implemented using def-use steps, then
+`clearsContent` might not be easy to use.
+
+## Type pruning
+
+The library supports pruning paths when a sequence of value-preserving steps
+originate in a node with one type, but reaches a node with another and
+incompatible type, thus making the path impossible.
+
+The type system for this is specified with the class `DataFlowType` and the
+compatibility relation `compatibleTypes(DataFlowType t1, DataFlowType t2)`.
+Using a singleton type as `DataFlowType` means that this feature is effectively
+disabled.
+
+It can be useful to use a simpler type system for pruning than whatever type
+system might come with the language, as collections of types that would
+otherwise be equivalent with respect to compatibility can then be represented
+as a single entity (this improves performance). As an example, Java uses erased
+types for this purpose and a single equivalence class for all numeric types.
+
+One also needs to define
+```
+Type Node::getType()
+Type Node::getTypeBound()
+DataFlowType getErasedRepr(Type t)
+string ppReprType(DataFlowType t)
```

Comment:
I realize that this is maybe not the place to discuss this, but I think we should change the interface to
```
DataFlowType getNodeType(Node n)
```
Then we can get rid of the three predicates above, in particular `getErasedRepr` which is currently exposed through the public `Node` class, although it is really an implementation detail.

Reply:
I have opened a PR for simplifying this here: https://github.com/github/codeql/pull/3854

Reply:
Cool :-)